rust/src/builtins/compose/mod.rs

```diff
@@ -6,74 +6,69 @@ pub(crate) mod commit;
-use openat_ext::OpenatDirExt;
-use std::ffi::{CStr, CString};
-use std::io::{self, Read};
+use rustix::fs::{FileType, MetadataExt};
+use std::io::Read;
+use std::os::unix::fs::DirBuilderExt;
+use std::os::unix::prelude::PermissionsExt;
 
 use crate::core::OSTREE_BOOTED;
 
 /// Prepare /dev and /run in the target root with the API devices.
 // TODO: delete this when we implement https://github.com/projectatomic/rpm-ostree/issues/729
 pub fn composeutil_legacy_prep_dev_and_run(rootfs_dfd: i32) -> CxxResult<()> {
-    let rootfs = crate::ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    let rootfs = unsafe { crate::ffiutil::ffi_dirfd(rootfs_dfd)? };
     legacy_prepare_dev(&rootfs)?;
-    rootfs.create_dir("run", 0o755)?;
-    rootfs.write_file(&OSTREE_BOOTED[1..], 0o755)?;
+    rootfs.create_dir_with("run", DirBuilder::new().mode(0o755))?;
+    rootfs.replace_contents_with_perms(&OSTREE_BOOTED[1..], b"", Permissions::from_mode(0o755))?;
     Ok(())
 }
 
 #[context("Preparing /dev hierarchy (legacy)")]
-fn legacy_prepare_dev(rootfs: &openat::Dir) -> Result<()> {
-    let src_dir = openat::Dir::open("/dev")?;
+fn legacy_prepare_dev(rootfs: &Dir) -> Result<()> {
+    let src_dir =
+        Dir::open_ambient_dir("/dev", cap_std::ambient_authority()).context("Opening host /dev")?;
     static CHARDEVS: &[&str] = &["full", "null", "random", "tty", "urandom", "zero"];
 
-    rootfs.ensure_dir("dev", 0o755)?;
-    rootfs.set_mode("dev", 0o755)?;
-    let dest_dir = rootfs.sub_dir("dev")?;
+    rootfs
+        .ensure_dir_with("dev", DirBuilder::new().mode(0o755))
+        .context("Creating dev")?;
+    rootfs
+        .set_permissions("dev", Permissions::from_mode(0o755))
+        .context("Setting permissions on target dev")?;
+    let dest_dir = &rootfs.open_dir("dev").context("Opening target dev")?;
 
-    for nodename in CHARDEVS {
-        let src_metadata = match src_dir.metadata_optional(*nodename)? {
+    for &nodename in CHARDEVS {
+        let src_metadata = match src_dir.metadata_optional(nodename).context(nodename)? {
             Some(m) => m,
             None => continue,
         };
-        // SAFETY: devnodes entries are plain non-NUL ASCII bytes.
-        let path_cstr = CString::new(*nodename).expect("unexpected NUL byte");
-        make_node_at(
-            &dest_dir,
-            &path_cstr,
-            src_metadata.stat().st_mode,
-            src_metadata.stat().st_rdev,
+        rustix::fs::mknodat(
+            dest_dir,
+            nodename,
+            FileType::CharacterDevice,
+            rustix::fs::Mode::from_bits_truncate(src_metadata.mode()),
+            src_metadata.rdev(),
         )
         .with_context(|| format!("Creating /dev/{}", nodename))?;
-        dest_dir.set_mode(*nodename, src_metadata.stat().st_mode)?;
+        dest_dir
+            .set_permissions(nodename, src_metadata.permissions())
+            .with_context(|| format!("Setting permissions of target {}", nodename))?;
```

Comment:
CI choked on this with:
```
error: Preparing /dev hierarchy (legacy): Setting permissions of target tty: No such device or address (os error 6)
```

But I'm not sure why `tty` is special here.

Reply:
Hmmm could it be that `/dev/tty` is a symlink?  It's not that way locally but...it seems possible.  I also notice that our openat-ext `set_mode()` explicitly no-ops symlinks, but `cap-std`'s `set_permissions` does not.

Reply:
Taking a step back, can we instead put all the devices/details in a table and directly use that?
I ported this logic from C to Rust as-is, and I was already puzzled at the time. Now (or after landing this) could be a good time to reconsider the whole approach.

Reply:
No objections to that if this keeps failing, but I would like to better understand the failure here because I think it will be part of our ongoing "lessons learned" from openat :arrow_right: cap-std porting.

As a cookie to myself (or anyone else who wants to switch approaches): https://github.com/systemd/systemd/blob/24759d8f089f7d8ecee460d92278af348af21dd1/src/core/namespace.c#L910 looks like the systemd code for this that we could adapt.

Reply:
It looks like it isn't a symlink:
```
crw-rw-rw-.  1 root root   5,   0 Mar 15 20:46 tty
[...]
error: Preparing /dev hierarchy (legacy): Setting permissions of target tty: No such device or address (os error 6)
```

Reply:
cap-std implements `set_permissions` by opening the file to obtain a file descriptor first, and doing `fchmod` on that, since Linux's `fchmodat` doesn't support `AT_NOFOLLOW_SYMLINK`, so cap-std can't prevent `fchmod` from following symlinks outside the sandbox. Is it possible this environment is configured such that a plain `chmod` or `fchmodat` on the "tty" file would work, but opening the "tty" file first doesn't work?

If so, I'm not sure what the right answer here is yet, but that's a separate question :-).

Reply:
Ah yes, that is likely to be it.  I think what's happening here is that `/dev/tty` may not actually be hooked up to anything, so opening the device will fail.

What's going on here is quite a complex chain of stuff; our CI runs as an unprivileged OpenShift pod but with `/dev/kvm` available, then we use the container userspace to assemble a VM userspace via `supermin` to do stuff that requires privileges, and this definitely a nonstandard environment.  I have lately been thinking we should toss out supermin for simply boot CoreOS with an Ignition config instead and run as a systemd unit.

But I don't understand why I can't reproduce this locally.

In this case though, since we created the device (and there can be no concurrent mutation by untrusted writers) there are no real concerns about following symlinks.

So...OK, pushed a bit to try out using `chmodat` directly.

Reply:
I wonder if opening the "tty" file with `O_PATH` would work.

Looking at cap-std's code, it does have code to use `O_PATH` but it likely isn't getting used here. I submitted https://github.com/bytecodealliance/cap-std/pull/238 which makes it try `O_PATH` first. I now think that's the right thing to do regardless, but there's also a chance it could fix the bug here too.

Of course, calling `chmodat` directly should also work :-).

Reply:
I tried an experiment with a deliberately invalid device file, and confirmed that `cap_std::fs::Dir::set_permissions` on it fails with the current cap-std, and succeeds with https://github.com/bytecodealliance/cap-std/pull/238. So it seems likely that that would fix the issue here too.